rfcs/shape-types.md

```diff
@@ -0,0 +1,87 @@
+# Shape types (concept name TBD)
+
+## Summary
+
+Introduce shape types, which are like table types, but are inhabited by both tables and class instances.
```

Comment:
It turns out that shape types are way more powerful than we initially thought. They are inhabited by anything that has a compatible _interface_.

Example 1:

```lua
-- f : <a : { lower: (a) -> b... }, b...>(s: a) -> b...
function f(s)
  return s:lower()
end

f("foo" :: string) -- ok
f("foo" :: "foo") -- ok
f("foo" :: "foo" | "bar") -- ok
f({ lower = function(self) return self.value end, value = "foo" }) -- ok
f({}) -- not ok
-- and so on...
```

This works out because instances of `string` has a `lower` method that came from their metatable's `__index` which points at the global `string` table, which provides the `lower` function, `(string) -> string`.

Example 2:

```lua
-- apply : <a : { metatable { __call: (b) -> c... } }, b, c...>(f: a, x: b) -> c...
function apply(f, x)
  return f(x)
end

apply(function(x) return x + 2 end, 2) -- ok
apply(setmetatable({}, { __call = function(x) return x + 2 end }), 2) -- ok
```

Example 3:

```lua
function weird(f)
  local x = f() -- constrains f to be some callable shape
  return f.something -- evidence that f cannot be an actual function
end
```

Type inference can flow easily into a table/class-only shape that provides `__call` and a property `something` instead of doing something very weird, such as a constrained type e.g. `(...) -> ... | { metatable { __call: (...) -> ... } }` and then finally deduce that it cannot be an actual function from `f.something`.

I think we get the point. It's no longer inhabitable by just tables or classes. It's inhabitable by anything that satisfies all the requirements of the said shape.

Reply:
Also worth noting this affects normalization as well as type refinements, e.g. 

```lua
local function f(s)
  local something = s:lower() -- t1 where t1 = { lower: (t1) -> a... }
  if typeof(s) == "string" then -- refine t1 & string into string for s
    -- s : string
  end
end
```